container/build_image.sh

```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/bash
+# Helper script to generate a karapace image.
+#
+# Notes:
+# - The script will always createa fresh temporary directory to run from. This
+# has a few benefits:
+#  - Ensures a clean state for copying files into the docker image. This is
+#  important specially important for the generate version.py file.
+#  - Makes sure the code in the container is not from a dirty working tree, so
+#  the commit in the project's version is the real code.
+#  - Prevents errors with concurrent changes, e.g. if you're generating a image
+#  while coding, and unintentionally save a file.
+#  - Effective uses of the caching layers.
+#
+# If for some reason you want to include the dirty state in the container just
+# call docker build directly.
+
+ARG_COMMIT=$1
+ARG_TAG_NAME=${2:-$1}
+
+COMMIT=$(git rev-parse -q --verify "${ARG_COMMIT}^{commit}")
+# replaces every occurence of / with -
+# this is useful if the commit is qualified with the repo name, e.g. aiven/master
+TAG_NAME=${ARG_TAG_NAME////-}
+
+if [[ -z "${COMMIT}" ]]; then
+    echo "Invalid commit provided '${COMMIT}'"
+    echo ""
+    echo "$0 <commitish_object> [:tag]"
+    exit 1
+fi
+
+code_checkout=$(mktemp --directory --suffix=-karapace-image)
+trap "rm -rf ${code_checkout}" EXIT
+
+git clone $(pwd) "${code_checkout}"
+
+pushd ${code_checkout}
+git checkout $COMMIT
+
+# HACK: Delete files from the `.git` which change on every operation.
+#
+# - .git/logs/HEAD - command history
+# - .git/index - binary file for the current index, very important for a
+#    working repository, not interesting for our image
+#
+# Being extra cautious here, to make sure only the temporary copy is modified.
+rm ${code_checkout}/.git/logs/HEAD
+rm ${code_checkout}/.git/index
```

Comment:
This is the most questionable change introduced by the PR. This improves cache reuse by removing the files that change when git is used, for instance, a git checkout will be logged into `.git/logs/HEAD` making the contents of the repo different for each run.

An alternative here would be to have this script run the `version.py` script outside of the container, and skip the copy of the repo into the container. I guess both work and I just didn't want to re implement this in the interest of time.

Reply:
Would it work to list those files in `.dockerignore` to solve caching etc. and skip not wanted files from ending up into container image?